src/main/kotlin/io/github/mojira/arisa/modules/RevokeConfirmationModule.kt

```diff
@@ -0,0 +1,45 @@
+package io.github.mojira.arisa.modules
+
+import arrow.core.Either
+import arrow.core.extensions.fx
+import net.rcarz.jiraclient.ChangeLogEntry
+
+data class RevokeConfirmationModuleRequest(
+    val confirmationStatus: String,
+    val changeLog: List<ChangeLogEntry>
+)
+
+class RevokeConfirmationModule(
+    private val getGroups: (String) -> Either<Throwable, List<String>>,
+    private val setConfirmationStatus: (String) -> Either<Throwable, Unit>,
+    private val confirmationFiledId: String
+) : Module<RevokeConfirmationModuleRequest> {
+    override fun invoke(request: RevokeConfirmationModuleRequest): Either<ModuleError, ModuleResponse> = with(request) {
+        Either.fx {
+            val confirmationChanges = changeLog.filter(::isConfirmationChange)
+            val volunteerChanges = confirmationChanges.filter(::changedByVolunteer)
+            val lastVolunteerChange = volunteerChanges.lastOrNull()
+            val volunteerConfirmation = lastVolunteerChange?.let { getConfirmation(it) } ?: "Unconfirmed"
+
+            assertNotEquals(confirmationStatus, volunteerConfirmation).bind()
+
+            setConfirmationStatus(volunteerConfirmation).toFailedModuleEither().bind()
+        }
+    }
+
+    private fun isConfirmationChange(entry: ChangeLogEntry) =
+        entry.items.any { it.field == confirmationFiledId }
+
+    private fun getConfirmation(entry: ChangeLogEntry) =
+        entry.items.lastOrNull { it.field == confirmationFiledId }?.toString
+
+    private fun changedByVolunteer(entry: ChangeLogEntry): Boolean {
+        val groups = getGroups(entry.author.name)
+
+        return if (groups.isLeft())
+        // when in doubt, assume change was done by a volunteer
+            true
```

Comment:
Not entirely sure what should be done here.
It's kinda annoying to have a filter method that can fail, and a case that we didn't have to deal with yet. It would maybe a bit better to somehow collect these errors, but the filter method can't easily deal with Eithers.

Reply:
You can do maps and filters inside eithers (they are monads) to combine them if needed

Reply:
I'm not quite sure how you mean that.
I don't see how combining a map with a filter would help here, or what Either has to do with that.

For this, it would be cool to have a filter in combination with an Either, so that the result is left when the filter function fails at some point for an entry of the list, but I don't see how to do that.
There might be other functional constructs I'm not aware of that are made for this case, of course.